Support for `libcurl` on Windows

jjerphan opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

[`Azure::Core::Http::CurlTransport`](https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/core/azure-core/inc/azure/core/http/curl_transport.hpp#L202) et al. are explicitly used by some project downstream (e.g. see [ArcticDB usages](https://github.com/search?q=repo%3Aman-group%2FArcticDB%20CurlTransport&type=code)).

This PR brings support for libcurl on Windows accordingly in addition of the current support of WinHTTP.

@conda-forge-admin, please rerender

jjerphan committed:
Support for libcurl on Windows

Signed-off-by: Julien Jerphanion <git@jjerphan.xyz>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

Commit:
MNT: Re-rendered with conda-build 25.1.2, conda-smithy 3.47.0, and conda-forge-pinning 2025.03.27.19.44.52

jjerphan committed:
Use CMake <4 for now

Confirmed that cmake is the culprit of failure by:
https://github.com/conda-forge/azure-core-cpp-feedstock/pull/25

Signed-off-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14127545367](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14127545367). Examine the logs at this URL for more detail.</sub>

jdblischak commented:
@jjerphan please ping us to review once the vc-feedstock issue has been resolved and the CI is green. Thanks!

jjerphan commented:
FYI, I won't be able to unlock the situation before a few days.

teo-tsirpanis commented on recipe/meta.yaml:
```diff
@@ -23,7 +23,7 @@ requirements:
+    - cmake <4
     - ninja
   host:
     - libcurl
```
If we decide to not do this, we should remove this dependency from Windows.

teo-tsirpanis reviewed (commented):
I'm moderately against using `CAPath` as the reason to enable the `libcurl` transport on Windows. Windows has a centralized system-managed store of trusted CA certificates, ~~and libcurl will ignore the custom CA path either way~~. [Turns out it does support it](https://curl.se/libcurl/c/CURLOPT_CAINFO.html#:~:text=(Schannel)%20This%20option%20is%20supported%20for%20Schannel%20in%20Windows%207%20or%20later), but I still think it's a bad idea to do from a security perspective.

I recommend ArcticDB (or its feedstock) to `#ifdef` away this code path on Windows, [like what TileDB does](https://github.com/TileDB-Inc/TileDB/blob/b171d128922ec2ba3c78193593cc40d1f51cafcc/tiledb/sm/filesystem/azure.cc#L1092-L1134). I would be open to this change if there is a more compelling use case.

jjerphan commented:
Thank you for your comment, @teo-tsirpanis. I was unaware of this centralized system-managed store of trusted CA certificates.

What do you think of:
 - repodata-patching the windows builds to remove the dependency of `libcurl` for all the Windows builds
 - removing `libcurl` as a dependency for the current build using WinHTTP as you propose in https://github.com/conda-forge/azure-core-cpp-feedstock/pull/23#discussion_r2019800491
 - introducing an alternative specification to come up with builds using `libcurl` only

?

teo-tsirpanis commented:
> repodata-patching the windows build to remove the dependency of libcurl for all the Windows builds

We can do it, but I'm not sure it is necessary to retroactively remove a tiny dependency that is otherwise commonly used by other packages.

> introducing an alternative specification to come up with builds using `libcurl` only

If a compelling use case for the `libcurl` transport on Windows arises, I would just enable both in the same build.

jjerphan commented:
ArcticDB uses `libcurl` for the Azure SDK because other SDKs and dependencies it relies on also use `libcurl`; hence my previous proposals above to be able to distinguish builds and keep the ones using WinHTTP separate from the one using `libcurl`.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you, but it looks like there was nothing to do.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14174044277](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14174044277). Examine the logs at this URL for more detail.</sub>

jjerphan commented:
Actually, ArcticDB needs both `libcurl` and `WinGet`.

I let @phoebusm give more background.

teo-tsirpanis commented:
> ArcticDB uses libcurl for the Azure SDK

In the [search results](https://github.com/search?q=repo%3Aman-group%2FArcticDB%20CurlTransport&type=code) you linked in the OP, the only use of `CurlTransport` by AcrticDB is to set `CAInfo` and `CAPath`, which should not be done on Windows in the first place. Hence my suggestion to add an `#ifndef _WIN32` around this code path. https://github.com/man-group/ArcticDB/blob/62dac1556fc12b4f564d9cf6da08f28eadd111af/cpp/arcticdb/storage/azure/azure_client_impl.cpp#L35-L44

jjerphan commented:
I am not the code owner, I let Phoebus reply.

phoebusm commented:
I think we can skip the `libcurl` dependency for azure sdk windows build for now. `ArcticDB` uses `winhttp` as backend for azure sdk, which is the default backend for windows.

jjerphan commented:
Removing `libcurl` as a dependency of `azure-core-cpp` in https://github.com/conda-forge/azure-core-cpp-feedstock/pull/26 then.
